internal/advtools/convo.lua

```diff
@@ -61,8 +61,111 @@ local function generateKeywordsForChoice(choice)
+end
+
+local function addIdentityWhereaboutsChoice(identity)
+    for i, c in pairs(adventure.conversation.conv_choice_info) do
+        if c.cc.type == df.talk_choice_type.AskWhereabouts and
+           c.cc.invocation_target_hfid ~= -1 and
+           c.cc.invocation_target_hfid == identity.impersonated_hf then
+            -- Don't add repeat entries
+            return
+        end
+    end
+    local identity_name = identity.name
+    local name = ""
+    local creature = df.creature_raw.find(identity.race)
+    if creature then
+        local caste = creature.caste[identity.caste]
+        name = caste.caste_name[0]
+    else
+        -- no race given for the identity, assume it's the histfig
+        local histfig = df.historical_figure.find(identity.histfig_id)
+        creature = df.creature_raw.find(histfig.race)
+        if creature then
+            local caste = creature.caste[histfig.caste]
+            name = caste.caste_name[0]
+        end
+    end
+    local title = "Ask for the whereabouts of the " .. name .. " " .. dfhack.TranslateName(identity_name, true)
+    local choice = new_choice(df.talk_choice_type.AskWhereabouts, title)
+    -- insert before the last choice, which is usually "back"
+    adventure.conversation.conv_choice_info:insert(#adventure.conversation.conv_choice_info-1, choice, dfhack.TranslateName(identity_name):split())
+    choice.cc.invocation_target_hfid = identity.impersonated_hf
+end
+
 -- Condense the rumor system choices
 local function rumorUpdate()
+    local conversation_state = adventure.conversation.conv_act.event[0].state
+    -- add new conversation options depending on state
+
+    -- If we're asking about directions, add ability to ask about all our relationships - visual, historical and identities.
+    -- In vanilla, we're only allowed to ask for directions to people we learned in anything that's added to df.global.adventure.rumor
+    if conversation_state == df.conversation_state_type.AskDirections then
+        local adventurer_figure = df.historical_figure.find(dfhack.world.getAdventurer().hist_figure_id)
+        local relationships = adventurer_figure.info.relationships
+
+        local visual = relationships.hf_visual
+        local historical = relationships.hf_historical
+        local identity = relationships.hf_identity
+
+        for _, profile in pairs(visual) do
+            addHistFigWhereaboutsChoice(profile)
+        end
+
+        -- This option will likely always fail unless the false identity is impersonating someone
+        -- but giving away the false identity's true historical figure feels cheap.
+        for _, profile in pairs(identity) do
+            addIdentityWhereaboutsChoice(df.identity.find(profile.identity_id))
+        end
+
+        -- Historical entities go last so as to not give away fake identities
+        for _, profile in pairs(historical) do
+            addHistFigWhereaboutsChoice(profile)
+        end
```

Comment:
with the loop through `adventure.conversation.conv_choice_info`, each of these has quadradic runtime complexity. perhaps it would be a good idea to make a single loop through `adventure.conversation.conv_choice_info` and cache the results? (and update the cache as you add more entries)

Reply:
also, `ipairs` is semantically appropriate for vectors, not `pairs`